File: Capítulo_5/Deque/Usando a classe Deque/Código em Python/deque.py
class Deque:
    def __init__(self):
        self.count = 0
        self.lowest_count = 0
        self.items = {}

    def addFront(self, element):
        if (self.is_empty()):
            self.addBack(element)
        elif (self.lowest_count > 0):
            self.lowest_count -= 1
            self.items[self.lowest_count] = element
        else:
            for i in range(self.count, 0 , -1):
                self.items[i] = self.items[i - 1]

            self.count += 1
            self.lowest_count = 0
            self.items[0] = element
        
    def addBack(self, element):
        self.items[self.count] = element
        self.count += 1

    def removeFront(self):
        if (self.is_empty()):
            return None
        result = self.items[self.lowest_count]
        del self.items[self.lowest_count]
        self.lowest_count += 1
        return result
    
    def removeBack(self):
        if (self.is_empty()):
            return None
        self.count -= 1
        result = self.items[self.count]
        del self.items[self.count]
        return result

    def peekFront(self):
        if (self.is_empty()):
            return None
        return self.items[self.lowest_count]
    
    def peekBack(self):
        if (self.is_empty()):
            return None
        return self.items[self.count]
    
    def size(self):
        return self.count - self.lowest_count
    
    def is_empty(self):
        return self.size() == 0
    
    def clear(self):
        self.items = {}
        self.count = 0
        self.lowest_count = 0
    
    def to_string(self):
        if (self.is_empty()):
            return ""
        obj_string = f"{self.items[self.lowest_count]}"
        for i in range(self.lowest_count + 1, self.count):
            obj_string = f"{obj_string}, {self.items[i]}"
        return obj_string
    
deque = Deque()
print(deque.is_empty()) # True

deque.addBack("John")
deque.addBack("Jack")
print(deque.to_string()) # John, Jack

deque.addBack("Camila")
print(deque.to_string()) # John, Jack, Camila
print(deque.size()) # 3
print(deque.is_empty()) # False
deque.removeFront() # remove John
print(deque.to_string()) # Jack, Camila
deque.removeBack() # Camila decide sair
print(deque.to_string()) # Jack
deque.addFront("John") # John retorna para pedir uma informação
print(deque.to_string()) # John, Jack

deque.clear()
print(deque.is_empty()) # True

File: Capítulo_5/Deque/Usando a classe Deque/deque.js
export default class Deque {
    constructor() {
        this.count = 0;
        this.lowestCount = 0;
        this.items = {};
    }

    addFront(element) {
        if (this.isEmpty()) {
            this.addBack(element);
        } else if (this.lowestCount > 0) {
            this.lowestCount--;
            this.items[this.lowestCount] = element;
        } else {
            for (let i = this.count; i > 0; i--) {
                this.items[i] = this.items[i - 1];
            }
            this.count++;
            this.lowestCount = 0;
            this.items[0] = element;
        }
    }

    addBack(element) {
        this.items[this.count] = element;
        this.count++;
    }

    removeFront() {
        if (this.isEmpty()) {
            return undefined;
        }
        const result = this.items[this.lowestCount];
        delete this.items[this.lowestCount];
        this.lowestCount++;
        return result;
    }

    removeBack() {
        if (this.isEmpty()) {
            return undefined;
        }
        this.count--;
        const result = this.items[this.count];
        delete this.items[this.count];
        return result;
    }

    peekFront() {
        if (this.isEmpty()) {
            return undefined;
        }
        return this.items[this.lowestCount];
    }

    peekBack() {
        if (this.isEmpty()) {
            return undefined;
        }
        return this.items[this.count - 1]
    }

    size() {
        return this.count - this.lowestCount;
    }

    isEmpty() {
        return this.size() === 0;
    }

    clear() {
        this.items = {};
        this.count = 0;
        this.lowestCount = 0;
    }

    toString() {
        if (this.isEmpty()) {
            return "";
        }
        let objString = `${this.items[this.lowestCount]}`;
        for (let i = this.lowestCount + 1; i < this.count; i++) {
            objString = `${objString}, ${this.items[i]}`;
        }
        return objString;
    }
}

// const deque = new Deque();
// console.log(deque.isEmpty()); // true

// deque.addBack("John");
// deque.addBack("Jack");
// console.log(deque.toString()); // John, Jack

// deque.addBack("Camila");
// console.log(deque.toString()); // John, Jack, Camila
// console.log(deque.size()); // 3
// console.log(deque.isEmpty()); // false
// deque.removeFront(); // remove John,
// console.log(deque.toString()); // Jack, Camila
// deque.removeBack();; // Camila decide sair
// console.log(deque.toString()); // Jack
// deque.addFront("John"); // John retorna para pedir uma informação
// console.log(deque.toString()); // John, Jack

// deque.clear();
// console.log(deque.isEmpty())// true